diff --git a/defaults/initrd.scripts b/defaults/initrd.scripts
index f8cc861deb716cdb6f12be0e549c13ea80794f0a..5d96aea96225554f85cf25d5ddda51cdab4f2606 100644
--- a/defaults/initrd.scripts
+++ b/defaults/initrd.scripts
@@ -1830,6 +1830,12 @@ openLUKS() {
 	local KEY_ERROR=0 KEYDEV_ERROR=0
 	local mntheader="/mnt/header/" mntkey="/mnt/key/" crypt_filter_ret=
 
+	if [ $2 -gt 0 ]
+	then
+		eval local LUKS_NAME="$1_$2"
+		eval local OPENED_LOCKFILE='"${CRYPT_'${TYPE}'_OPENED_LOCKFILE}_$2"'
+	fi
+
 	if [ -z "${LUKS_DEVICE}" ]
 	then
 		bad_msg "'crypt_${1}' kernel command-line argument is not set!"
@@ -2505,7 +2511,11 @@ start_LUKS_root() {
 
 	if [ -n "${CRYPT_ROOT}" ]
 	then
-		openLUKS "root"
+		local crypt_index=0
+		for CRYPT_ROOT in ${CRYPT_ROOTS}; do
+			openLUKS "root" ${crypt_index}
+			let crypt_index=${crypt_index}+1
+		done
 		if [ -n "${REAL_ROOT}" ]
 		then
 			# Rescan volumes
@@ -2535,7 +2545,11 @@ start_LUKS_swap() {
 
 	if [ -n "${CRYPT_SWAP}" ]
 	then
-		openLUKS "swap"
+		local crypt_index=0
+		for CRYPT_SWAP in ${CRYPT_SWAPS}; do
+			openLUKS "swap" ${crypt_index}
+			let crypt_index=${crypt_index}+1
+		done
 		if [ -z "${REAL_RESUME}" ]
 		then
 			# Resume from swap as default
@@ -2569,9 +2583,11 @@ start_sshd() {
 	write_env_file \
 		"${CRYPT_ENV_FILE}" \
 		CRYPT_ROOT \
+		CRYPT_ROOTS \
 		CRYPT_ROOT_OPTIONS \
 		CRYPT_SILENT \
 		CRYPT_SWAP \
+		CRYPT_SWAPS \
 		CRYPT_SWAP_OPTIONS
 
 	run touch /var/log/lastlog
